### DISCUSSION TITLE
Approximation and uncertainty via reification

### DISCUSSION THREAD

**Discussion by rybesh** [periodo/edtf-ontology#8] [Ideas]
EDTF Levels 1 and 2 introduce qualification operators (`?`, `~` and `%`) which can be used to qualify dates or parts of dates as uncertain, approximate, or both uncertain and approximate. In Level 1 these operators can only qualify entire date expressions, while in Level 2 they can qualify parts of date expressions.

This kind of qualification seems to be precisely the kind of thing that the [RDF reification vocabulary](https://www.w3.org/TR/rdf-schema/#ch_reificationvocab) was developed for. Reification means making RDF statements about RDF statements. In this case, we are creating an RDF statement describing a datetime, and then creating another RDF statement that qualifies the former statement as uncertain, approximate, or both.

See the following cases for examples:

* [qualified calendar date](https://github.com/periodo/edtf-ontology/blob/main/cases/level-1/qualification/day-precision/)
* [qualified reduced precision date (year and month only)](https://github.com/periodo/edtf-ontology/blob/main/cases/level-1/qualification/month-precision/)
* [qualified reduced precision date (year only)](https://github.com/periodo/edtf-ontology/blob/main/cases/level-1/qualification/year-precision/)
* [qualified date component(s)](https://github.com/periodo/edtf-ontology/blob/main/cases/level-2/qualification/individual/)
* [qualified date component group(s)](https://github.com/periodo/edtf-ontology/blob/main/cases/level-2/qualification/group/)

The draft EDTF Ontology defines two classes that can be used to type instances of [`rdf:Statement`](https://www.w3.org/TR/rdf-schema/#ch_statement) as approximate or uncertain:

* [`edtfo:ApproximateStatement`](https://periodo.github.io/edtf-ontology/#Approximatestatement)
* [`edtfo:UncertainStatement`](https://periodo.github.io/edtf-ontology/#Uncertainstatement)

Both of these classes are subclasses of [`edtfo:QualifiedStatement`](https://periodo.github.io/edtf-ontology/#Qualifiedstatement) to make it possible to query for all qualified statements regardless of whether they are uncertain or approximate or both.

An advantage of this approach is that qualified datetimes result in exactly the same triples as unqualified datetimes, so they can be queried without concern for qualification. When one wants to filter out or otherwise treat qualified statements differently, the separate qualification statements can be used.

Some disadvantages of this approach are that
* it is verbose, and
* it requires that both the subject and the object of the qualified statement be identified (i.e. they cannot be blank / anonymous nodes).

Fortunately, the new RDF-star specification seems to address both of these issues. 

If I understand correctly, the following triples describing an EDTF expression for “approximately June 2004”:

```turtle
:when a time:Instant ;
  # Year-month approximate.
  edtfo:inEDTFDateTime "2004-06~" ;
  time:inDateTime :2004_06
  .

:2004_06 a time:DateTimeDescription ;
  time:unitType time:unitMonth ;
  time:year 2004 ;
  time:month 6
  .

:temporalPositionStatement a edtfo:ApproximateStatement ;
  rdf:subject :when ;
  rdf:predicate time:inDateTime ;
  rdf:object :2004_06
  .
```

…could be more succinctly expressed using RDF-star as follows:

```turtle
:when a time:Instant ;
  # Year-month approximate.
  edtfo:inEDTFDateTime "2004-06~" ;
  time:inDateTime [
    a time:DateTimeDescription ;
    time:unitType time:unitMonth ;
    time:year 2004 ;
    time:month 6
  ] {| a edtfo:ApproximateStatement |} .
```

**Comment by nichtich** [comment]
RDF Reification and RDF-star are not the same and they come with additional complexity. Some alternative modeling possibilities for `2004-06~`:

New properties `edtfo:approximate` and `edtfo:uncertain`:

~~~ttl
:when a time:Instant ;
  edtfo:inEDTFDateTime "2004-06~" ;
  time:inDateTime [
    a time:DateTimeDescription ;
    time:unitType time:unitMonth ;
    time:year "2004"^^xsd:gYear ;
    time:month "--06"^^xsd:gMonth 
    edtfo:approximate true
  ]
~~~

Subclass of `time:DateTimeDescription`:

~~~ttl
:when a time:Instant ;
  edtfo:inEDTFDateTime "2004-06~" ;
  time:inDateTime [
    a edtfo:ApproximateDateTimeDescription ;
    time:unitType time:unitMonth ;
    time:year "2004"^^xsd:gYear ;
    time:month "--06"^^xsd:gMonth 
  ]
~~~

The question is what use case do you assume? I assume a simple query "give me all events from 2004-06" should return approximate and uncertain dates as well, unless explicitly excluded in the query. So I'd prefer one or two additional properties to explicitly mark instances of [time:GeneralDateTimeDescription](https://www.w3.org/TR/owl-time/#time:GeneralDateTimeDescription) as approximate and/or uncertain.
